### ISSUE TITLE
Add scopes to autograd profiler

### ISSUE THREAD

**Issue by d4l3k** [pytorch/pytorch#28249]
## 🚀 Feature

Add scopes within a torchscript model so you can get profile information per scope for both the forward and backward pass.

## Motivation

For distributed execution (model parallelism) we want to be able to measure the exact CPU time spent on different parts of a torchscript module so we can shard the model in a reasonable manner to different machines. We have high level "components" that represent shardable units of execution and want to use the autograd profiler to profile per component.

## Pitch

We want to add two custom methods into torch so we can describe the scope. Making them in C++ allows us to use them in torchscript as well in python.

TorchScript pseudocode
```py
def forward(self, x):
  handle = torch._profiler_scope_enter("foo")
  x = torch.relu(x)
  ...
  torch._profiler_scope_exit(handle)
  return x
```

Python Sugar
```py
def forward(self, x):
  with profiler.scope("foo"):
    x = torch.relu(x)
    ...
    return x
```

These methods would add scope information to RecordFunction and thus provide it to the autograd Profiler. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/record_function.h

The current RECORD_FUNCTION implementation appears to track the lineage from backwards functions to the forward pass via sequence_nr so it should be easy to walk the RecordFunction tree and extract scope information. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.h#L104-L116 https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L238

One notable thing is that sequence_nr is currently thread_local. From what I've heard, the autograd/backward pass can be multithreaded and thus we will likely need to extend sequence_nr to include a thread ID as well as the current thread_local sequence_nr.

## Alternatives

There's ways to do this when running in the python environment w/ register_backward_hook + a custom nn.Module however there's no equivalent way to do this with torchscript since it doesn't support register_backward_hook. The autograd backward pass can also run multithreaded from my understanding so that would break that approach anyways.

**Comment by d4l3k** [comment]
cc @dzhulgakov @ngimel @aazzolini @bddppq

**Comment by bddppq** [comment]
cc @ilia-cher

**Comment by bddppq** [comment]
> the autograd/backward pass can be multithreaded

Actually even if it's not multi-threaded, since autograd has its own thread-pool, the backward ops will be invoked by a thread different from the one that executed the forward op (which is also the one that created and assigned its thread local sequence_nr to the backward op). I think we can either 1. encode the thread id into the thread local sequence_nr (like using first two bytes to store the thread id at initialization time) or 2. add an additional field (say "creation_thread_id") to the `torch::autograd::Node` class. Both should have very minimal perf impact (but we should run benchmark to proof it).

**Comment by albanD** [comment]
@bddppq Why do we want to change that logic? Why would that help here?

**Comment by ilia-cher** [comment]
I believe we already have very similar functionality (albeit it might not be available from TorchScript). Let me find a link to some previous discussion..

**Comment by ilia-cher** [comment]
Some relevant discussions in https://github.com/pytorch/pytorch/issues/19422

**Comment by ilia-cher** [comment]
After https://github.com/pytorch/pytorch/pull/23428 we could make `with torch.autograd.profiler.record_function("scope_name")` traceable/scriptable

**Comment by d4l3k** [comment]
@albanD since the sequence IDs are multithreaded you need some way to reliably tie the backward pass to the forward pass scopes.

@ilia-cher good to know, I'll take a look at doing that

**Comment by bddppq** [comment]
@ilia-cher @d4l3k Oh I didn't know the existence of `torch.autograd._push_range` (and `torch.autograd._pop_range`). But looks like they are directly operating at the `profiler::Event` level, I think we should probably change them to create (and destroy) a RecordFunction (and call its `.before` and dtor to trigger the profiler's push and pop range). The benefit is then it will work for other registered Callback. Re. jit support, I think the easiest way to implement is to register the RecordFunction creation and destroy functions as operators.

**Comment by bddppq** [comment]
@albanD Yeah what @d4l3k said (basically addressing this https://github.com/pytorch/pytorch/blob/ce16d68/torch/autograd/profiler.py#L423-L432)
